### PR TITLE
Relayer: Transaction confirmation logic

### PR DESCRIPTION
**Pull request by sczembor**
## Description

Closes: #40 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
    
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary

### CODE REVIEW

File: nbtc/relayer.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+				//TODO: add proper error handling here and decide which errors we shutdown the relayer
+				if err.Error() == "error updating transaction status" {
+					log.Err(err).Msg("Critical error updating transaction status, shutting down")
+					close(r.shutdownChan)
+				} else {
+					log.Err(err).Msg("Error processing transactions, continuing")
+				}
+			}
+		case <-ticker2.C:
+			r.checkConfirmations()
```
Comment (sczembor) [inline nbtc/relayer.go:82]: @robert-zaremba @vuvoth Here both the `processPendingTxs` and `checkConfirmations` will be operating on the same database. Is there a risk for a race con? Should use mutex on the db? Please let me know what you guys think

File: dal/daltest/helpers.go [hunk 2]
```diff
@@ -17,3 +18,42 @@ func InitTestDB(t *testing.T) *dal.DB {
 	assert.NilError(t, err)
 	return db
 }
+
+// GetHashBytes creates a byte array from a hash string.
+func GetHashBytes(t *testing.T, hashString string) []byte {
```
Comment (vuvoth) [inline dal/daltest/helpers.go:23]: If we don't use this function outside package, I think we don't need to export this. Let's make this a private function.

File: dal/db.go [hunk 3]
```diff
@@ -64,15 +66,15 @@ func (db DB) InitDB() error {
 
 // InsertTx inserts a new transaction into the database
 func (db DB) InsertTx(tx Tx) error {
```
Comment (vuvoth) [inline dal/db.go:68]: Should we use *DB, because we modify the db data?

File: nbtc/relayer.go [hunk 4]
```diff
@@ -20,10 +21,17 @@ type Relayer struct {
...
-func NewRelayer(btcClientConfig rpcclient.ConnConfig, processTxsInterval time.Duration, db *dal.DB) (*Relayer, error) {
+func NewRelayer(
+	btcClientConfig rpcclient.ConnConfig,
+	processTxsInterval time.Duration,
+	confirmTxsInterval time.Duration,
+	db *dal.DB,
+) (*Relayer, error) {
+
 	if db == nil {
 		return nil, fmt.Errorf("database cannot be nil")
```
Comment (vuvoth) [inline nbtc/relayer.go:36]: I think we should use logger more than `fmt`.

File: nbtc/relayer.go [hunk 5]
```diff
@@ -43,17 +51,21 @@ func NewRelayer(btcClientConfig rpcclient.ConnConfig, processTxsInterval time.Du
 		processTxsInterval = time.Second * 5
 	}
 
+	if confirmTxsInterval == 0 {
+		confirmTxsInterval = time.Second * 7
```
Comment (vuvoth) [inline nbtc/relayer.go:55]: Can you explain what this is?

Comment (sczembor) [reply nbtc/relayer.go:55]: So basically if the value passed is 0, we set it by deafult to 7 seconds. This is how often the relayer will query the node to update the broadcasted transactions status

File: dal/daltest/helpers.go (hunk 2 above)
Comment (sczembor) [reply dal/daltest/helpers.go:23]: we actually use it in different packages. eg. dal_test

File: dal/db.go (hunk 3 above)
Comment (sczembor) [reply dal/db.go:68]: You are right. Updated it

File: nbtc/relayer.go (hunk 4 above)
Comment (sczembor) [reply nbtc/relayer.go:36]: so should i add logger? We dont want to just log the problem and conitune because the relayer without the db has no point

Comment (sczembor) [reply nbtc/relayer.go]: I added the logging of the error but still return the error here

File: nbtc/relayer.go [hunk 6]
```diff
@@ -102,6 +128,36 @@ func (r *Relayer) processPendingTxs() error {
...
+		if err != nil {
+			return err
+		}
+		txDetails, err := r.btcClient.GetTransaction(hash)
+		if err != nil {
+			return fmt.Errorf("error getting transaction details: %w", err)
+		}
+
+		// TODO: decide what threshold to use. Read that 6 is used on most of the cex'es etc.
+		if txDetails.Confirmations >= 6 {
```
Comment (vuvoth) [inline nbtc/relayer.go:150]: Can we set confirmation constant in parameters?

File: nbtc/relayer.go [hunk 7]
```diff
@@ -20,12 +21,21 @@ type Relayer struct {
 	btcClient        bitcoin.Client
 	shutdownChan     chan struct{}
 	processTxsTicker *time.Ticker
+	confirmTxsTicker *time.Ticker
 }
 
 // NewRelayer creates a new Relayer instance with the given configuration.
-func NewRelayer(btcClientConfig rpcclient.ConnConfig, processTxsInterval time.Duration, db *dal.DB) (*Relayer, error) {
+func NewRelayer(
+	btcClientConfig rpcclient.ConnConfig,
```
Comment (vuvoth) [inline nbtc/relayer.go:29]: Group those parameters into 2 groups: config and db. Do you think this is more readable and maintainable?

File: nbtc/relayer.go [hunk 8]
```diff
@@ -20,12 +21,21 @@ type Relayer struct {
...
+) (*Relayer, error) {
+
 	if db == nil {
-		return nil, fmt.Errorf("database cannot be nil")
+		err := fmt.Errorf("database cannot be nil")
+		log.Err(err).Msg("")
+		return nil, err
 	}
 
 	if btcClientConfig.Host == "" || btcClientConfig.User == "" || btcClientConfig.Pass == "" {
```
Comment (vuvoth) [inline nbtc/relayer.go:41]: We can call this the default config and create the function `DefaultConfig()` to return this struct.

File: nbtc/relayer.go (hunk 7 above)
Comment (vuvoth) [reply nbtc/relayer.go:29]: Feel free to do this in other PR.

**Review by vuvoth — APPROVED** [review]

File: nbtc/relayer.go (hunk 6 above)
Comment (sczembor) [reply nbtc/relayer.go]: Done

File: nbtc/relayer.go (hunk 7 above)
Comment (sczembor) [reply nbtc/relayer.go:37]: I grouped it in two different configs, leaving the db seperate

File: nbtc/relayer.go (hunk 8 above)
Comment (sczembor) [reply nbtc/relayer.go:48]: I am not sure this is neccesary since we will be using it once

Comment (sczembor) [reply nbtc/relayer.go:48]: + we already have the defult values in the constructor

**Pull request merged**